src/controller/python/chip/ChipStack.py

```diff
@@ -367,7 +393,25 @@ def Call(self, callFunct, timeoutMs: int = None):
             return self.callbackRes
         return res
 
-    def CallAsync(self, callFunct):
+    async def CallAsync(self, callFunct, timeoutMs: int = None):
+        '''Run a Python function on CHIP stack, and wait for the response.
+        This function will post a task on CHIP mainloop and waits for the call response in a asyncio friendly manner.
+        '''
+        loop = asyncio.get_event_loop()
+        future = loop.create_future()
+        callObj = AsyncSimpleCallableHandle(callFunct, loop, future)
+        pythonapi.Py_IncRef(py_object(callObj))
+
+        res = self._ChipStackLib.pychip_DeviceController_PostTaskOnChipThread(
+            self.cbHandleChipThreadRun, py_object(callObj))
+
+        if not res.is_success:
+            pythonapi.Py_DecRef(py_object(callObj))
+            raise res.to_exception()
+
+        return await asyncio.wait_for(future, timeoutMs / 1000 if timeoutMs else None)
+
+    def CallAsyncWithCallback(self, callFunct):
```

Comment:
> After spending way to long trying to figure out the difference between `Call` and `CallAsync` (old, not your version) and noticing a lot of dead code in ChipStack.py that makes it seems like they are functionally relatively the same it is the ugly reaching in of `self._ChipStack.callbackRes = ...` in `ChipDeviceCtrl.py` that ultimately drives `CallAsync` forward. I see why you went with `CallAsyncWithCallback`, but we need to document that method better, what do we mean `WithCallback`

Actually, no, that is not the reason I had in mind, but I see why you came to that conclusion.The reason is closer: The call to `blockingCB`.  In the end, this shows the downside of asynchronous programming without asyncio: You end up in callback hell :see_no_evil: . So maybe `CallAsyncWithCallbacks` (note the s :sweat_smile: ). 

I like to rewrite all that code to asyncio. It seems all not to hard, I start to get a grasp with the code. I just hope that all the tests aren't too much work to rewrite. I started to rewrite some code in https://github.com/project-chip/connectedhomeip/pull/32497/commits, but I need to revisit that PR.

As for the naming: `CallAsyncWithSelfSuppliedReleaseCallback` doesnt' sound too bad to me. It is scary enough for everyone trying to use it to think twice about the implications :sweat_smile:

Reply:
>The reason is closer: The call to `blockingCB`

But even then, the only thing that actually sets that blockingCB is `src/controller/python/chip/ChipBluezMgr.py`. All the other calls, to the currently named `CallbackAsync`, essentially need `completeEvent` to be set otherwise it will be stuck in that look forever.

If you want to use `CallAsyncWithCallbacks` that is fine by me as long as we update the description to reflect that the caller needs to be consider if they want to set `blockingCB`, and that in order to have a return that they need to have some callback handling mechanism to set `completeEvent`

> I like to rewrite all that code to asyncio. 

I fully support this, because what we have currently is hard to understand and very unclean.

Reply:
> But even then, the only thing that actually sets that blockingCB is `src/controller/python/chip/ChipBluezMgr.py`. All the other calls, to the currently named `CallbackAsync`, essentially need `completeEvent` to be set otherwise it will be stuck in that look forever.

Yeah agreed, the `completeEvent` is actually more the reason what makes this function unique.

I renamed it to `CallAsyncWithCompleteCallback`.

> I fully support this, because what we have currently is hard to understand and very unclean.

Cool, thanks, good to here. The only question will be if we can/should make "outside" APIs async too :thinking: . E.g. `Commission`, `CommissionOnNetwork` and friends, can we just convert them to co-routines?

Reply:
> Cool, thanks, good to here. The only question will be if we can/should make "outside" APIs async too 🤔 . E.g. `Commission`, `CommissionOnNetwork` and friends, can we just convert them to co-routines?

If you have the time to explore that I will review any PRs. Free to DM me the PRs.

With the updated comment. This is resolved from my perspective